### PR TITLE
Add support for Constraint Programming

### PR DESCRIPTION
**Pull request by odow**
Part of https://github.com/jump-dev/MathOptInterface.jl/issues/1805

### CODE REVIEW

**Comment by odow** [comment]
I've marked this as DNMY, because it is the base of a stack of PRs for constraint programming. We can merge the others into here, and then wait a while before merging this into MOI proper.

File: src/sets.jl [hunk 1]
```diff
@@ -1116,6 +1116,33 @@ struct Complements <: AbstractVectorSet
     end
 end
 
+"""
+    AllDifferent(dimension::Int)
+
+Constrain the elements of a vector-valued function so that no two elements take
+the same value.
```
Comment (odow) [inline src/sets.jl:1123]:
```suggestion
the same value and each element is integer.
```

File: src/sets.jl [hunk 2]
```diff
@@ -1143,6 +1143,361 @@ struct AllDifferent <: AbstractVectorSet
...
+        end
+        return new(dimension)
+    end
+end
+
+"""
+    Path(from::Vector{Int}, to::Vector{Int})
+
+Given a graph comprised of a set of nodes `1..N` and a set of arcs `1..E`
+represented by an edge from node `from[i]` to node `to[i]`, `Path` constraints
```
Comment (blegat) [inline src/sets.jl:1455]:
```suggestion
represented by an edge from node `from[i]` to node `to[i]`, `Path` constrains
```

File: src/sets.jl [hunk 3]
```diff
@@ -1143,6 +1143,361 @@ struct AllDifferent <: AbstractVectorSet
...
+end
+
+function Base.:(==)(x::BinPacking{T}, y::BinPacking{T}) where {T}
+    return x.capacity == y.capacity && x.weights == y.weights
+end
+
+"""
+    Cumulative(dimension::Int)
+
+The set ``\\{(s, d, r, b) \\in \\mathbb{Z}^{length(s)+length(d)+length(r)+1}\\}``
```
Comment (blegat) [inline src/sets.jl:1361]:
```suggestion
The set ``\\{(s, d, r, b) \\in \\mathbb{R}^{length(s)+length(d)+length(r)+1}\\}``
```

Comment (blegat) [reply src/sets.jl:1361]: I would imagine that MIP formulations would support these to be any real numbers

File: src/sets.jl [hunk 4]
```diff
@@ -1143,6 +1143,361 @@ struct AllDifferent <: AbstractVectorSet
...
+
+"""
+    Path(from::Vector{Int}, to::Vector{Int})
+
+Given a graph comprised of a set of nodes `1..N` and a set of arcs `1..E`
+represented by an edge from node `from[i]` to node `to[i]`, `Path` constraints
+the set
+``(s, t, ns, es) \\in (1..N)\\times(1..N)\\times\\{0,1\\}^N\\times\\{0,1\\}^E``,
+to form subgraph that is a path from node `s` to node `t`, where node `n` is in
+the path if `ns[n]` is `1`, and edge `e` is in the path if `es[e]` is `1`.
```
[the remark this answers is not in the document]
Comment (odow) [reply src/sets.jl:1462]: It must be acyclic and traverse each selected node and edge exactly once.

File: src/sets.jl [hunk 5]
```diff
@@ -1143,6 +1143,361 @@ struct AllDifferent <: AbstractVectorSet
...
+
+Base.copy(set::Table) = Table(copy(set.table))
+
+Base.:(==)(x::Table{T}, y::Table{T}) where {T} = x.table == y.table
+
+"""
+    Circuit(dimension::Int)
+
+The set ``\\{x \\in \\{1..d\\}^d\\}`` that constraints ``x`` to be a circuit,
+such that ``x_i = j`` means that ``j`` is the successor of ``i``.
```
[the remark this answers is not in the document]
Comment (odow) [reply src/sets.jl:1428]: No. Clarified this in the new docstring.

File: src/sets.jl (hunk 3 above)
Comment (odow) [reply src/sets.jl]: I think most users and solvers of constraint programming in MOI are going to use `Model{Int}`, so I guess this becomes a moot point.

**Comment by odow** [comment]
@blegat okay to merge? Then I'll start on the easy bridges.

**Comment by odow** [comment]
Merging because I think we're pretty happy. We can always make changes in follow-up PRs.

**Pull request merged**